File: cognix-flask-server/utils/cron_service.py
"""
Cron Job Service for Flask Server
Keeps both Node.js and Flask servers alive by pinging health endpoints
"""

import requests
import time
import os
from datetime import datetime
from threading import Thread
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Server URLs from environment variables
NODE_SERVER_URL = os.getenv('NODE_SERVER_URL', 'http://localhost:3000')
FLASK_SERVER_URL = os.getenv('FLASK_SERVER_URL', 'http://localhost:5000')
PING_INTERVAL = int(os.getenv('PING_INTERVAL', '10'))  # seconds


def ping_server(server_name: str, url: str) -> bool:
    """
    Ping a server's health endpoint
    
    Args:
        server_name: Name of the server
        url: Health check URL
        
    Returns:
        bool: True if ping successful, False otherwise
    """
    try:
        response = requests.get(
            url,
            timeout=5,
            headers={'User-Agent': 'Cognix-Cron-Job/1.0'}
        )
        
        timestamp = datetime.now().isoformat()
        print(f"✅ [{timestamp}] {server_name} is alive - Status: {response.status_code}")
        return True
        
    except Exception as e:
        timestamp = datetime.now().isoformat()
        print(f"❌ [{timestamp}] {server_name} ping failed: {str(e)}")
        return False


def ping_node_server():
    """Continuously ping Node.js server"""
    url = f"{NODE_SERVER_URL}/health"
    while True:
        ping_server('Node.js Server', url)
        time.sleep(PING_INTERVAL)


def ping_flask_server():
    """Continuously ping Flask server"""
    url = f"{FLASK_SERVER_URL}/health"
    while True:
        ping_server('Flask Server', url)
        time.sleep(PING_INTERVAL)


def initialize_cron_jobs():
    """Initialize and start cron jobs in separate threads"""
    print('\n' + '=' * 60)
    print('🕐 Initializing Cron Jobs for Server Health Checks')
    print('=' * 60)
    print(f'📍 Node Server: {NODE_SERVER_URL}/health')
    print(f'📍 Flask Server: {FLASK_SERVER_URL}/health')
    print(f'⏰ Ping Interval: {PING_INTERVAL} seconds')
    print('=' * 60 + '\n')
    
    # Create daemon threads for continuous pinging
    node_thread = Thread(target=ping_node_server, daemon=True)
    flask_thread = Thread(target=ping_flask_server, daemon=True)
    
    # Start threads
    node_thread.start()
    flask_thread.start()
    
    print('✅ Cron jobs started successfully!')
    print(f'⏰ Pinging both servers every {PING_INTERVAL} seconds...\n')


if __name__ == '__main__':
    try:
        initialize_cron_jobs()
        
        # Keep the main thread alive
        while True:
            time.sleep(1)
            
    except KeyboardInterrupt:
        print('\n👋 Stopping cron jobs...')

File: cognix-server/services/cronService.js
/**
 * Cron Job Service for Cognix
 * Keeps both Node.js and Flask servers alive by pinging health endpoints
 */

const cron = require('node-cron');
const axios = require('axios');
require('dotenv').config();

// Server URLs from environment variables
const NODE_SERVER_URL = process.env.NODE_SERVER_URL || 'http://localhost:3000';
const FLASK_SERVER_URL = process.env.FLASK_SERVER_URL || 'http://localhost:5000';

/**
 * Ping a server's health endpoint
 * @param {string} serverName - Name of the server
 * @param {string} url - Health check URL
 */
async function pingServer(serverName, url) {
  try {
    const response = await axios.get(url, {
      timeout: 5000, // 5 second timeout
      headers: {
        'User-Agent': 'Cognix-Cron-Job/1.0'
      }
    });
    
    console.log(`✅ [${new Date().toISOString()}] ${serverName} is alive - Status: ${response.status}`);
    return true;
  } catch (error) {
    console.error(`❌ [${new Date().toISOString()}] ${serverName} ping failed:`, error.message);
    return false;
  }
}

/**
 * Initialize cron jobs
 */
function initializeCronJobs() {
  console.log('\n' + '='.repeat(60));
  console.log('🕐 Initializing Cron Jobs for Server Health Checks');
  console.log('='.repeat(60));
  console.log(`📍 Node Server: ${NODE_SERVER_URL}/health`);
  console.log(`📍 Flask Server: ${FLASK_SERVER_URL}/health`);
  console.log('='.repeat(60) + '\n');

  // Ping Node.js server every 10 seconds
  cron.schedule('*/10 * * * * *', async () => {
    await pingServer('Node.js Server', `${NODE_SERVER_URL}/health`);
  });

  // Ping Flask server every 10 seconds
  cron.schedule('*/10 * * * * *', async () => {
    await pingServer('Flask Server', `${FLASK_SERVER_URL}/health`);
  });

  console.log('✅ Cron jobs started successfully!');
  console.log('⏰ Pinging both servers every 10 seconds...\n');
}

// Start cron jobs if this file is run directly
if (require.main === module) {
  initializeCronJobs();
  
  // Keep the process running
  process.on('SIGINT', () => {
    console.log('\n👋 Stopping cron jobs...');
    process.exit(0);
  });
}

module.exports = { initializeCronJobs, pingServer };
